### PR TITLE
Update use of allPolicies variable and push notification subscription to fix inconsistency in deep link routing

### PR DESCRIPTION
**Pull request by tienifr**
### Details

- Currently, logging out from an account with workspaces and then login to a new account will not open the FAB modal as default. To make the FAB modal always opened for a new account regardless of the previous state of the app, delete the duplicated `allPolicies` variable in https://github.com/Expensify/App/blob/df900108b413826dd7c4b0e51ab6cf8092223289/src/libs/actions/Welcome.js#L68-L78
and reuse the same variable in https://github.com/Expensify/App/blob/21486cd0173aca0f3ac2d9991c9898063af8fb54/src/libs/actions/Policy.js#L18
- Currently, navigating to a deeplink from push notification may cause inconsistency due to overlapping navigation with the default navigation (last accessed report) of the app. To make the push notification always open the correct report, add the use of `Navigation.isDrawerReady()` before navigating in https://github.com/Expensify/App/blob/a59825caa19d1e3217e8d59db864dbe2f3bd386d/src/libs/actions/Report.js#L72-L82
- Currently, the device push notification listener is not working when user logout, keep app in background and login again. When such scenario happens, clicking a push notification will open the app but will not open the correct report, which also cause inconsistency. To miltigate this issue, add the subscription logic in the `register()` function like in https://github.com/Expensify/App/pull/14588/files#diff-24c7a3c4dd0b7b7cb20e0e8cb485e27259b55286e6395fc6ae92727c7bb8741f
This function is then called each time the user try to login to the app, but the logic will only be triggered for _new_ user logging in the app only (logged in users are not affected)

### Fixed Issues

$ https://github.com/Expensify/App/issues/13691
https://github.com/Expensify/App/issues/13691#issuecomment-1384218222
https://github.com/Expensify/App/issues/13691#issuecomment-1384218945

### Tests
1. Open the app
2. Register with a new email
3. Go back to home screen
4. Perform 1 of the following 
  4.1. Keep the app in background
  4.2. Kill the app in background
5. Go to mailbox and click the magic sign-in link to access the set password page. Set a password and proceed
6. Verify that the FAB modal is opened after signing in
7. Select _New workspace_ to create a new workspace
8. Wait for the Notification from the support team member
9. Perform 1 of the following 
  9.1. Keep the app in background
  9.2. Kill the app in background
10. Tap the notification
11. Verify that the app navigates to the correct channel #admins

- [x] Verify that no errors appear in the JS console

### Offline tests

1. Open the app
2. Register with a new email
3. Go back to home screen
4. Perform 1 of the following 
  4.1. Keep the app in background
  4.2. Kill the app in background
5. Go to mailbox and click the magic sign-in link to access the set password page. Set a password and proceed
6. Disable the internet
7. Verify that the FAB modal is opened after signing in
8. Enable the internet
9. Select _New workspace_ to create a new workspace
10. Wait for the Notification from the support team member
11. Disable the internet
12. Perform 1 of the following 
  12.1. Keep the app in background
  12.2. Kill the app in background
13. Tap the notification
14. Verify that the app navigates to the correct channel #admins

### QA Steps

1. Open the app
2. Register with a new email
3. Go back to home screen
4. Perform 1 of the following 
  4.1. Keep the app in background
  4.2. Kill the app in background
5. Go to mailbox and click the magic sign-in link to access the set password page. Set a password and proceed
6. Verify that the FAB modal is opened after signing in
7. Select _New workspace_ to create a new workspace
8. Wait for the Notification from the support team member
9. Perform 1 of the following 
  9.1. Keep the app in background
  9.2. Kill the app in background
10. Tap the notification
11. Verify that the app navigates to the correct channel #admins

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/113963320/214908780-5ae73c5a-297f-4610-8125-e7f865c5097b.mp4

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/113963320/215100351-5ff4cbaa-3d8d-40a6-b692-7506044b03ea.mp4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/113963320/214905840-e19a485a-1033-4daf-b789-371ff050176f.mp4

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/113963320/215014902-09da2142-2a5c-4b7a-95cf-ddc148f0583d.mp4

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/113963320/214866143-512c619d-100c-405d-a7ba-c81d6bdaf683.mov

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/113963320/214831559-a38bdddc-d2c1-4e76-b09b-1c787783719c.mp4

</details>

### CODE REVIEW

**Comment by mountiny** [comment]
@tienifr please make sure to link the issue correctly when the PR is marked as raedy for a reivew, so the correct people are assigned for a review

image


@ctkochan22 assigning myself for a review since I helped as CME on that linked issue.

**Comment by tienifr** [comment]
@mountiny sorry for the inconvenience, I'll note that for next time
thanks!

**Comment by ctkochan22** [comment]
@tienifr can you also update the PR title so that it explains what is being changed/fixed?

Its hard to keep track now, but clear titles will be helpful in the future when debugging changes and git blames in the future

**Comment by mountiny** [comment]
@tienifr aand also some short details section explaining you changes in plain English is also great if someone comes to this PR some time in future 🙌 Thanks!

**Comment by tienifr** [comment]
@mountiny @ctkochan22 PR title and details are updated!

File: src/libs/actions/Welcome.js [hunk 1]
```diff
@@ -65,15 +65,14 @@ Onyx.connect({
...
-const allPolicies = {};
 Onyx.connect({
     key: ONYXKEYS.COLLECTION.POLICY,
     callback: (val, key) => {
         if (!val || !key) {
             return;
         }
 
-        allPolicies[key] = {...allPolicies[key], ...val};
+        Policy.allPolicies[key] = {...Policy.allPolicies[key], ...val};
```
[the remark this answers is not in the document]
Comment (tienifr) [reply src/libs/actions/Welcome.js:75]: @rushatgabhane I'm sorrry, but did you mean that you prefer **Solution 1** for the `2) Magic sign in Failure issue`, in which I added the clearing logic for the variable? You said in the issue comment that you prefer the `delete allPolicies` solution, which I believe was more similar to **Solution 2**

**Comment by tienifr** [comment]
> > Currently, the device push notification listener is not working when user logout, keep app in background and login again. When such scenario happens, clicking a push notification will open the app but will not open the correct report, which also cause inconsistency. To miltigate this issue, add the subscription logic in the register() function like in
> 
> I'm sorry, I'm lost here. Is there an issue for this? Why are we fixing push notifications in this PR? I have no context sweat_smile

@rushatgabhane 
Yes, it is the `3) Admin chat failure` issue, quote `It should be the #admins chat, but is often an entirely different chat`. When the push notification event listener is not working, clicking the push notification after creating a workspace (or any other types of notification like chat noti) will not navigate the user to the right chat, causing the inconsistency. The issue occurred when an user logged out of an account (event listener got cleared), and then login right again to another account (event listener not reinitialized).

File: src/libs/actions/Welcome.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (tienifr) [reply src/libs/actions/Welcome.js]: @rushatgabhane New commit with **Solution 1** for the `Magic sign in Failure issue` has been uploaded

https://github.com/Expensify/App/pull/14588/commits/408f9726f151308b2cca6f6c657cb0aa7a1ff82c

[the remark this answers is not in the document]
Comment (tienifr) [reply src/libs/actions/Welcome.js]: @rushatgabhane I have updated a new commit, and posted a comment to explain your concert below

**Review by ctkochan22 — DISMISSED** [review]

File: src/libs/actions/Policy.js [hunk 2]
```diff
@@ -1027,4 +1027,5 @@ export {
     openWorkspaceMembersPage,
     openWorkspaceInvitePage,
     removeWorkspace,
+    allPolicies,
```
[the remark this answers is not in the document]
Comment (tienifr) [reply src/libs/actions/Policy.js]: @rushatgabhane Yes indeed, that's a nice catch. I have updated it in the latest commit.
https://github.com/Expensify/App/commit/f2a52fc6adefc309114eaec6a0ff19c7ca9cea54

**Review by rushatgabhane — COMMENTED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/29673073/215469523-0477e2c1-bf1f-40c2-8cd0-b34a19bc6f64.mov


</details>

<details>
<summary>Mobile Web - Chrome</summary>


https://user-images.githubusercontent.com/29673073/215553594-85ac178d-bd41-4fff-b87e-69f991fcb81d.mp4



</details>

<details>
<summary>Mobile Web - Safari</summary>


https://user-images.githubusercontent.com/29673073/215552395-add3aa44-3865-40d0-958c-374d79e8bc75.mov



</details>


<details>
<summary>Desktop</summary>



https://user-images.githubusercontent.com/29673073/215551945-876b7802-f8b3-4176-b362-566d99ae4358.mov



</details>

<details>
<summary>iOS</summary>


https://user-images.githubusercontent.com/29673073/215551409-df7244d1-448f-4eda-8259-2eb39bc8a1ed.mov



</details>

<details>
<summary>Android</summary>


https://user-images.githubusercontent.com/29673073/215548577-7807b7d7-7281-4bce-a911-7efbe2b5ad84.mp4


</details>

**Comment by tienifr** [comment]
> Hey on DEV, I'm not getting any push notifications on iOS and Android. Asked on slack if that's just me https://expensify.slack.com/archives/C01GTK53T8Q/p1675092451431619

@rushatgabhane 
Hi, I encountered the same issue as well. For now, replacing the debug urban airship config with production will make it work

```diff
- appKey = uulSSfTDQJ2r0PMpjRrhmQ
- appSecret = D4Bhf0HrQEehrPua74Tyiw
+ appKey = 55vypj0ARc6cN09MX7ogtQ
+ appSecret = EsSaqbdLSvmyC6kSBFJCtQ
inProduction = false
developmentLogLevel = VERBOSE

# Notification Customization
notificationIcon = ic_notification
notificationAccentColor = #2EAAE2
```

**Comment by mountiny** [comment]
Removing my review request here since Jules is back

**Review by rushatgabhane — APPROVED** [review]
@Julesssss LGTM! 

@tienifr Thanks for the speedy help on getting the notification sorted ⚡

**Review by ctkochan22 — APPROVED** [review]

**Review by mountiny — APPROVED** [review]
👍 all yours @Julesssss

**Review by Julesssss — APPROVED** [review]

**Comment by Julesssss** [comment]
There's a new warning [showing](https://expensify.slack.com/archives/C01GTK53T8Q/p1675186823689109) on iOS against these changes. I don't think this is a regression, but it would be great to clear this up: 

Simulator Screen Shot - iPhone 14 Pro Max - 2023-01-31 at 23 00 57
Simulator Screen Shot - iPhone 14 Pro Max - 2023-01-31 at 23 00 45

**Comment by Julesssss** [comment]
actually, I’ll just fix this quickly
